src/crush/CrushWrapper.cc

```diff
@@ -1817,6 +1821,144 @@ int CrushWrapper::update_device_class(int id,
   return 1;
 }
 
+// Recursively get all only-child shadow parents up to the root
+// based on the given item id.
+// Return -EBUSY if any parent has more than one children.
+// Return -ENOENT if a specific parent can not be found.
+// Return 0 on success.
+int CrushWrapper::get_immediate_only_child_shadow_parents(CephContext *cct,
+                                                          int id,
+                                                          list<int> *parents)
+{
+  assert(parents);
+  parents->clear();
+  while (_search_item_exists(id)) {
+    int parent;
+    int r = get_immediate_parent_id(id, &parent, PARENT_SHADOW);
+    if (r < 0) {
+      ldout(cct, 0) << __func__ << " unable to get immediate shadow parent of"
+                    << " item id = " << id << dendl;
+      return -ENOENT;
+    }
+    crush_bucket *b = get_bucket(parent);
+    if (IS_ERR(b)) {
+      ldout(cct, 0) << __func__ << " unable to locate parent id = "
+                    << parent << dendl;
+      return -ENOENT;
+    }
+    if (b->size != 1) {
+      assert(b->size > 0);
+      ldout(cct, 0) << __func__ << " parent id = " << parent
+                    << " have " << b->size << " children,"
+                    << " returning" << dendl;
+      return -EBUSY;
+    }
+    parents->push_back(parent);
+    id = parent;
+  }
+  return 0;
+}
+
+int CrushWrapper::remove_device_class(CephContext *cct, int id, ostream *ss)
+{
```

Comment:
This won't work for the following case(multiple roots share the same class):

./bin/ceph osd tree up --show-shadow
```

ID  CLASS WEIGHT  TYPE NAME                                            UP/DOWN REWEIGHT PRI-AFF 
-10   ssd 2.00000 root default~ssd                                                              
 -9   ssd 2.00000     host gitbuilder-ceph-rpm-centos7-amd64-basic~ssd                          
  1   ssd 1.00000         osd.1                                             up  1.00000 1.00000 
  2   ssd 1.00000         osd.2                                             up  1.00000 1.00000 
 -8   ssd 1.00000 root foo~ssd                                                                  
 -4   ssd 1.00000     rack foo-rack~ssd                                                         
 -3   ssd 1.00000         host foo-host~ssd                                                     
  0   ssd 1.00000             osd.0                                         up  1.00000 1.00000 
 -7       1.00000 root foo                                                                      
 -6       1.00000     rack foo-rack                                                             
 -5       1.00000         host foo-host                                                         
  0   ssd 1.00000             osd.0                                         up  1.00000 1.00000 
 -1       2.00000 root default                                                                  
 -2       2.00000     host gitbuilder-ceph-rpm-centos7-amd64-basic                              
  1   ssd 1.00000         osd.1                                             up  1.00000 1.00000 
  2   ssd 1.00000         osd.2                                             up  1.00000 1.00000 
```

BTW, can we get https://github.com/ceph/ceph/pull/16577 into luminous too? It's really annoying if there is no obvious way to find out all the shadow trees, especially for debugging.